src/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+
+using Microsoft.CodeAnalysis;
+
+namespace System.Management.Automation.Internal.Generators
+{
+    /// <summary>
+    /// Source Code Generator to create partial PSVersionInfo class.
+    /// </summary>
+    [Generator]
+    public class PSVersionInfoGenerator : ISourceGenerator
+    {
+        /// <summary>
+        /// Generate output PSVersionInfo.generated.cs file.
+        /// This allows to directly get ProductVersion and others without reflection.
+        /// </summary>
+        /// <param name="context">Generator execution context.</param>
+        public void Execute(GeneratorExecutionContext context)
+        {
+            var result = CreatePSVersionInfoPartialClass(context);
+
+            // We must use specific file name suffix (*.g.cs,*.g, *.i.cs, *.generated.cs, *.designer.cs)
+            // so that Roslyn analyzers skip the file.
+            context.AddSource("PSVersionInfo.generated.cs", result);
+        }
+
+        /// <summary>
+        /// Not used.
+        /// </summary>
+        /// <param name="context">Generator initialization context.</param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            // No initialization required for this one.
+        }
+
+        /// <summary>
+        /// Create a string with partial PSVersionInfo class.
+        /// </summary>
+        /// <param name="context">Generator execution context.</param>
+        /// <returns>A string with partial PSVersionInfo class.</returns>
+        private static string CreatePSVersionInfoPartialClass(GeneratorExecutionContext context)
+        {
+            // We must put "<auto-generated" on first line so that Roslyng analizers skip the file.
+            const string SourceTemplate = @"
+// <auto-generated>
+// This file is auto-generated by PSVersionInfoGenerator.
+// </auto-generated>
+
+    namespace System.Management.Automation
+{{
+    public static partial class PSVersionInfo
+    {{
+        // Defined in PowerShell.Common.props as ProductVersion
+        // '6.0.0-beta.7 Commits: 29 SHA: 52c6b...'
+        internal static string ProductVersion {{ get; }} = ""{0}"";
+
+        // Defined in PowerShell.Common.props as
+        // git describe --abbrev=60 --long
+        // 'v6.0.0-beta.7-29-g52c6b...'
+        private static readonly string _rawGitCommitId = ""{1}"";
+
+        // Defined in PowerShell.Common.props as PSCoreBuildVersion
+        // '6.0.0-beta.7'
+        private static readonly string _mainVersion = ""{2}"";
+    }}
+}}";
```

Comment:
Notice, the file is written to disk and it is accessible for compliance tools.

@PaulHigin Friendly ping.

Reply:
@TravisEz13 , @iSazonov 
The concern, of course, is code injection.  I don't immediately see an issue, but this needs to be discussed in the next WG-Security meeting.  However, a meeting may not occur until December, due to the holidays.  Since this is an optional change, I feel it is not high priority.

Reply:
To be clear, this is the opinion of the @WG-Security group, not just me.  We looked at `const string SourceTemplate` where the above property values are concatenated into the source template as strings.  This is a supply chain source code injection opportunity, that would be difficult to detect.  Our understanding is that using a CodeDom is the preferred way to build up source code for compilation.

Reply:
@PaulHigin I really don't understand about CodeDom. The best comment from .Net team I saw about using CodeDom in SGs is https://github.com/dotnet/roslyn/issues/48214#issuecomment-704432433

Reply:
I'm not an expert either, and I'll defer to @TravisEz13.

Reply:
For code injection to happen here, there would need to be a double quote, right?  Can we just reject any strings that contain a double quote?

Reply:
I really don't understand. We could implement this as TypeGen and ResGen which we already use. Only difference is the TypeGen and ResGen are our custom utilities but the source generator is .Net feature.
To do injection, someone would have to compromise the entire repository in order to corrupt both build scripts and the source code.

Reply:
It's not clear if there is really a code injection vulnerability. The version properties are defined in `PowerShell.Common.props`, and then passed in by using the .NET source generator property `CompilerVisibleProperty`, so unless the properties are already tampered in `PowerShell.Common.props`, there seems no other way to cause a security problem, in my personal understanding.

Be noted that, today, the `ProductVersion` embedded in PowerShell assemblies is calculated using the same properties in `PowerShell.Common.props`. Then [in the static constructor](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/PSVersionInfo.cs#L77-L109) of `PSVersionInfo`, we get the `ProductVersion` from reflection, and parse the string to get PowerShell version and `GitCommitId`. So, if those properties could be tampered, then it's already a problem to our code today.

@TravisEz13 it would be very helpful if you can elaborate how the code injection could happen, and how would it affect our release builds or daily builds. Thanks!

Reply:
Comment resolved by https://github.com/PowerShell/PowerShell/pull/15603#issuecomment-1273819725